### PR TITLE
Create using_gpu_sanitizer.md

### PR DESCRIPTION
**Pull request by Rmalavally**
Draft documentation for using GPU sanitizer

### CODE REVIEW

**Review by Rmalavally — COMMENTED** [review]
Looks good. Let's add this as a guideline in contributing.md, so developers are aware. Thanks so much, Sam.

File: docs/understand/using_gpu_sanitizer.md [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+This document provides documentation on using ROCm Address Sanitizer.
+For information about LLVM Address Sanitizer, see https://clang.llvm.org/docs/AddressSanitizer.html.
+
+Compiling Address Sanitizer
+The address sanitizer process begins by compiling the application of interest with the address sanitizer instrumentation.
+
+Recommendations for doing this are:
+
+Compile as many application and dependent library sources as possible using an AMD-built clang-based compiler such as amdclang++.
```
Comment (samjwu) [inline docs/understand/using_gpu_sanitizer.md:15]:
```suggestion
Compile as many application and dependent library sources as possible using an AMD-built clang-based compiler such as `amdclang++`.
```

File: docs/understand/using_gpu_sanitizer.md [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+Add the following options to the existing compiler and linker options:
+`-fsanitize=address` - enables instrumentation)
+`-shared-libsan` - use shared version of runtime)
+`-g` - add debug info for improved reporting)
+
+Explicitly use `xnack+` in the offload architecture option. For example, `--offload-arch=gfx90a:xnack+`
+Other architectures are allowed, but their device code will not be instrumented and a warning will be emitted.
+
+It is not an error to compile some files without address sanitizer instrumentation, but doing so reduces the ability of the process to detect addressing errors. However, if the main program "a.out" does not directly depend on the Address Sanitizer runtime (libclang_rt.asan-x86_64.so) after the build completes (check by running ldd or readelf), the application will immediately report an error at runtime as described in the next section.
```
Comment (samjwu) [inline docs/understand/using_gpu_sanitizer.md:25]:
```suggestion
It is not an error to compile some files without address sanitizer instrumentation, but doing so reduces the ability of the process to detect addressing errors. However, if the main program "a.out" does not directly depend on the Address Sanitizer runtime (`libclang_rt.asan-x86_64.so`) after the build completes (check by running `ldd` or `readelf`), the application will immediately report an error at runtime as described in the next section.
```

File: docs/understand/using_gpu_sanitizer.md [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+About Compilation Time
+When `-fsanitize=address`` is used, the LLVM compiler adds instrumentation code around every memory operation. This added code must be handled by all of the downstream components of the compiler toolchain and results in increased overall complilation time. This increase is especially evident in the AMDGPU device compiler and has in a few instances raised the compile time to an unacceptable level.
```
Comment (samjwu) [inline docs/understand/using_gpu_sanitizer.md:28]:
```suggestion
When `-fsanitize=address` is used, the LLVM compiler adds instrumentation code around every memory operation. This added code must be handled by all of the downstream components of the compiler toolchain and results in increased overall compilation time. This increase is especially evident in the AMDGPU device compiler and has in a few instances raised the compile time to an unacceptable level.
```

File: docs/understand/using_gpu_sanitizer.md [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+There are a few options if the compile time becomes unacceptable:
+
+- Avoid instrumentation of the files which have the worst compile times. This will reduce the effectiveness of the address sanitizer process.
+- Add the option `-fsanitize-recover=address`` to the compiles with the worst compile times. This option simplifies the added instrumentation resulting in faster compilation. See below for more information.
```
Comment (samjwu) [inline docs/understand/using_gpu_sanitizer.md:33]:
```suggestion
- Add the option `-fsanitize-recover=address` to the compiles with the worst compile times. This option simplifies the added instrumentation resulting in faster compilation. See below for more information.
```

File: docs/understand/using_gpu_sanitizer.md (hunk 3 above)
Comment (Rmalavally) [reply docs/understand/using_gpu_sanitizer.md:28]: This is fixed.

File: docs/understand/using_gpu_sanitizer.md (hunk 4 above)
Comment (Rmalavally) [reply docs/understand/using_gpu_sanitizer.md:33]: Fixed

**Review by samjwu — APPROVED** [review]

**Pull request merged**